File: frontend/src/types/seatAllocation.ts
import type { Seat } from "./seat";

export type AllocationStatus = "active" | "released" | "cancelled";

export type SeatAllocation = {
  id: number;
  employee_id: number;
  seat_id: number;
  project_id: number;
  status: AllocationStatus;
  allocated_at: string;
  released_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
};

export type SeatAllocationCreate = {
  employee_id: number;
  seat_id: number;
  project_id?: number | null;
  notes?: string | null;
};

export type SeatAllocationRelease = {
  allocation_id?: number | null;
  employee_id?: number | null;
  seat_id?: number | null;
  notes?: string | null;
};

export type SeatAllocationListParams = {
  page?: number;
  page_size?: number;
  employee_id?: number;
  seat_id?: number;
  project_id?: number;
  status?: AllocationStatus;
};

export type SeatSuggestion = {
  seat: Pick<Seat, "id" | "floor" | "zone" | "bay" | "seat_number" | "status">;
  score: number;
  reason: string;
};

File: backend/app/schemas/seat_allocation.py
"""Seat allocation Pydantic schemas."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import AllocationStatus
from app.schemas.common import PaginatedResponse


class SeatAllocationCreate(BaseModel):
    """Payload for assigning a seat to an employee."""

    employee_id: int
    seat_id: int
    project_id: int | None = Field(
        None,
        description="Optional override; defaults to the employee's current project.",
    )
    notes: str | None = None


class SeatAllocationRelease(BaseModel):
    """Payload for releasing a seat allocation."""

    allocation_id: int | None = None
    employee_id: int | None = None
    seat_id: int | None = None
    notes: str | None = None


class SeatAllocationRead(BaseModel):
    """Seat allocation response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    employee_id: int
    seat_id: int
    project_id: int
    status: AllocationStatus
    allocated_at: datetime
    released_at: datetime | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class SeatSuggestion(BaseModel):
    """Suggested seat with a simple proximity score."""

    seat: dict
    score: int
    reason: str


class SeatAllocationListResponse(PaginatedResponse[SeatAllocationRead]):
    """Paginated allocation list."""
